agent/drive_handler.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+const (
+	blockPath = "/sys/block"
+	drivePath = "/dev"
+
+	blockMajorMinor = "dev"
+	blockSize       = "size"
+)
+
+type drive struct {
+	Name       string
+	DriveID    string
+	MajorMinor string
+	Size       int
+	DrivePath  string
+}
+
+type driveHandler struct {
+	// drives is a mapping to all the stub drives
+	drives map[string]drive
+	// blockPath is the path which lists all block devices. In linux this is
+	// typically /sys/block
+	BlockPath string
+	// drivePath is the directory where the list of drives will be found. In
+	// linux this is typically /dev
+	DrivePath string
+}
+
+func newDriveHandler(blockPath, drivePath string) *driveHandler {
+	return &driveHandler{
+		drives:    map[string]drive{},
+		BlockPath: blockPath,
+		DrivePath: drivePath,
+	}
+}
+
+func (dh driveHandler) GetDrive(id string) (drive, bool) {
+	v, ok := dh.drives[id]
+	return v, ok
+}
+
+// DiscoverDrives will iterate the block path in the sys directory to retrieve all
+// stub block devices.
+func (dh *driveHandler) DiscoverDrives() error {
```

Comment:
We should probably use a mutex to write lock `DiscoverDrives` so that if `GetDrive` is ever called at the same time it will wait for the map to be fully populated before returning

Reply:
I can fix that by simply moving `DiscoverDrives` into the constructor and making the method unexported. Would much prefer that than a mutex.